### PR TITLE
Use 2PC in adaptive executor when dealing with replication factors above 1

### PR DESCRIPTION
**Pull request by serprex**
DESCRIPTION: Use 2PC in adaptive executor when dealing with replication factors above 1

While this passes `check-multi`, with `check-base EXTRA_TESTS="multi_modifications"` we end up in a state where both shards are healthy, yet worker1 has kept the insert & worker2 is missing it

May be relevant to #2823

### CODE REVIEW

**Comment by serprex** [comment]
I've updated this PR with a fix

File: src/test/regress/expected/multi_modifications_9.out [hunk 1]
```diff
@@ -356,6 +356,20 @@ ERROR:  relation "public.limit_orders_750000" does not exist
 -- Second: Move aside limit_orders shard on the second worker node
 ALTER TABLE renamed_orders RENAME TO limit_orders_750000;
 -- Connect back to master node
```
Comment (pykello) [inline src/test/regress/expected/multi_modifications_9.out]: Move the comment

**Comment by marcocitus** [comment]
I'm not entirely against this change (see #1032), but it could have substantial performance impact on users with replication factor >1.

**Comment by onderkalaci** [comment]
Marking this as related to #2460, and actually fixes the part related to replication factor > 1.

> I'm not entirely against this change (see #1032), but it could have substantial performance impact on users with replication factor >1.

A quick benchmark on my local machine showed that single statement insert throughput dropped %50.

**Comment by serprex** [comment]
@marcocitus alternative is marking shards unhealthy. Feel 1PC should do that regardless of whether we keep this choice or not

**Comment by onderkalaci** [comment]
>@marcocitus alternative is marking shards unhealthy. Feel 1PC should do that regardless of whether we keep this choice or not

@marcocitus @serprex I'd actually vote for doing that, if there is a bug we should fix it. This change might hurt some users really badly and there is no turn back from this via a GUC etc.

**Pull request merged**